File: front-end/src/types/preference.ts
// A preference is a free-form statement from a user, optionally scoped to
// one room. Unscoped (roomId null) means "applies to every room this user
// is currently assigned to" — resolved dynamically, not stored as a list.
export interface PreferenceRule {
  id: string
  userId: string
  roomId: string | null
  roomName: string | null
  text: string
  enabled: boolean
  createdAt: string
}

export type PreferenceRuleInput = { userId: string; roomId: string | null; text: string; enabled: boolean }

File: back-end/smart-home-manager/src/main/java/com/tw/medtech/pfa/model/Preference.java
package com.tw.medtech.pfa.model;

import jakarta.persistence.*;
import lombok.*;
import lombok.experimental.SuperBuilder;
import org.hibernate.annotations.CreationTimestamp;

import java.time.Instant;

@Entity
@Table(name = "preferences")
@Getter
@Setter
@NoArgsConstructor
@SuperBuilder
public class Preference {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @ManyToOne
    @JoinColumn(name = "user_id", nullable = false)
    private User user;

    // Null = applies to every room this user is currently assigned to.
    @ManyToOne
    @JoinColumn(name = "room_id")
    private Room room;

    // Free-form — this is the actual preference, in the user's own words.
    @Column(length = 2000, nullable = false)
    private String text;

    @Column(nullable = false)
    @Builder.Default
    private boolean enabled = true;

    @CreationTimestamp
    private Instant createdAt;
}
